api/src/main/java/org/openmrs/module/fgh/mpi/MpiHttpClient.java

```diff
@@ -117,22 +105,98 @@ public List<Object> submitBundle(String bundleData) throws Exception {
 	 * @throws Exception
 	 */
 	public void submitPatient(String patientData) throws Exception {
+		MpiContext mpiContext = MpiContext.initIfNecessary();
+		
 		log.info("Submitting patient record to the MPI");
 		if (log.isDebugEnabled()) {
 			log.debug("Patient data -> " + patientData);
 		}
 		
-		List<Map<String, Object>> mpiIdsResp = submitRequest(SUBPATH_PATIENT, patientData, List.class);
-		
-		if (log.isDebugEnabled()) {
-			log.debug("MPI patient submission response: " + mpiIdsResp);
+		if (mpiContext.getMpiSystem().isSanteMPI()) {
+			Map<String, Object> santeResponse = submitRequest(SUBPATH_PATIENT, patientData, Map.class);
+			
+			if (log.isDebugEnabled()) {
+				log.debug("MPI patient submission response: " + santeResponse);
+			}
+			
+			if (MapUtils.isEmpty(santeResponse)) {
+				throw new APIException("An empty response was received when the patient was submitted");
+			}
+		} else {
+			List<Map<String, Object>> mpiIdsResp = submitRequest(SUBPATH_PATIENT, patientData, List.class);
+			
+			if (log.isDebugEnabled()) {
+				log.debug("MPI patient submission response: " + mpiIdsResp);
+			}
+			
+			if (CollectionUtils.isEmpty(mpiIdsResp) || mpiIdsResp.get(0) == null) {
+				throw new APIException("An empty response was received when the patient was submitted");
+			}
 		}
 		
-		if (CollectionUtils.isEmpty(mpiIdsResp) || mpiIdsResp.get(0) == null) {
-			throw new APIException("An empty response was received when the patient was submitted");
+		log.info("Successfully submitted the patient record to the MPI");
+	}
+	
+	private <T> void retriveAccessToken(Class<T> responseType) throws Exception {
+		MpiContext mpiContext = MpiContext.initIfNecessary();
+		
+		String data = "grant_type=client_credentials" + "&" + "scope=*" + "&" + "client_secret="
+		        + mpiContext.getClientSecret() + "&" + "client_id=" + mpiContext.getClientId();
+		
+		// Request a Refresh token in case it expires 
+		if (mpiContext.getTokenInfo() != null) {
+			
+			if (!mpiContext.getTokenInfo().isValid(LocalDateTime.now())) {
+				//Implement a refresh token method
+				data = "grant_type=refresh_token&refresh_token=" + mpiContext.getTokenInfo().getRefreshToken() + "&"
+				        + "client_secret=" + mpiContext.getClientSecret() + "&" + "client_id=" + mpiContext.getClientId();
+				this.doAuthentication(data);
+			}
+		} else {
+			// Normal Login 
+			this.doAuthentication(data);
 		}
+	}
+	
+	protected void doAuthentication(String data) throws Exception {
+		MpiContext mpiContext = MpiContext.initIfNecessary();
 		
-		log.info("Successfully submitted the patient record to the MPI");
+		String uri = "/auth/oauth2_token";
+		String url = mpiContext.getServerBaseUrl() + uri;
+		
+		HttpURLConnection connection = MpiUtils.openConnection(url);
+		
+		Range<Integer> successRange = Range.between(200, 299);
+		
+		try {
+			connection.setRequestProperty("Content-type", "application/x-www-form-urlencoded");
```

Comment:
I believe we should have delegated  to `submitRequest` instead of duplicating this http client code here, I will create a ticket to address this in another ticket.

Reply:
Ok